### PR TITLE
feat(recipe): per-Builder DataProvider isolation; deprecate process globals

### LINKED ISSUE #985: feat(recipe): expose effective-values and manifest helpers on RecipeResult
Part of #983. Unblocks the BOM-per-leaf prereq in #966.

## Summary

The only way to obtain a component's resolved Helm values today is through `pkg/bundler.DefaultBundler.Make`, which writes files to disk. The merge logic itself — base values from `ValuesFile` deep-merged with `ComponentRef.Overrides` — lives in `mergeValues` in `pkg/recipe/adapter.go`, which is package-private.

Expose it as a public method on `*recipe.RecipeResult` so consumers (the BOM tool, in-process Go callers, future tooling) can obtain effective values without going through the bundler or duplicating the merge algorithm.

## Motivation

Two concrete consumers are blocked:

1. **#966's BOM rewrite.** Prereq 1 requires "resolve each through `pkg/recipe`, and for every `componentRef`, run `helm template` with the same effective values that bundling uses." There is no public API for that today. The BOM tool either reimplements values merging (drift hazard — two divergent implementations of the same merge order) or duplicates bundler internals (coupling hazard). Both are bad.

2. **In-process Go consumers.** Any tool that wants to inspect "what values would actually be deployed for component X under recipe Y" — diff tools, dry-run validators, GitOps generators — needs this primitive. Without it, every such consumer reinvents the merge.

The merge logic is correct and well-tested via the bundler. Promoting it to the public surface is a low-risk additive change; no new behavior, just a new entry point.

## Proposed change

Add two public methods on `*RecipeResult` (in `pkg/recipe/metadata.go` or a new `pkg/recipe/values.go`):

```go
// GetValuesForComponent returns the effective Helm values for the named
// component using the process-global DataProvider. Equivalent to the
// values that the bundler would write to values.yaml for this component.
//
// Merge order matches the bundler:
//   1. base values from ComponentRef.ValuesFile (if set)
//   2. overlay values (if the overlay system contributed any for this component)
//   3. ComponentRef.Overrides (highest precedence)
//
// Returns ErrCodeNotFound if name is not in r.ComponentRefs.
// Returns ErrCodeInternal on values-file read or parse failure.
func (r *RecipeResult) GetValuesForComponent(name string) (map[string]any, error)
```

And a manifest-content helper at package level for components that carry `ManifestFiles`:

```go
// GetManifestContent reads a manifest path (as recorded in
// ComponentRef.ManifestFiles or PreManifestFiles) from the process-
// global DataProvider and returns its contents.
//
// Returns ErrCodeNotFound when the file is absent.
// Returns ErrCodeInternal on read failure.
func GetManifestContent(path string) ([]byte, error)
```

### Semantics

- Both helpers MUST use the same merge / read logic the bundler currently uses internally. Refactor — do not reimplement. The bundler should be updated to call these new public helpers so we don't have two code paths to maintain.
- The returned map from `GetValuesForComponent` is owned by the caller and safe to mutate. Internal cached state must be deep-copied (see the related cache-corruption fix in the parent epic).
- A missing `ValuesFile` is not an error — it produces a values map derived purely from `Overrides`. An empty result is `map[string]any{}`, not nil, so callers can distinguish "no values" from a marshaling edge case.
- The package-private `mergeValues` stays private. The public contract is the `GetValuesForComponent` method, not the algorithm.

### Tests

- Add table-driven coverage in `pkg/recipe/metadata_test.go` (or `values_test.go`) for:
  - Component not in `ComponentRefs` → `ErrCodeNotFound`.
  - Component with no `ValuesFile`, no `Overrides` → empty map.
  - Component with `ValuesFile` only → values from file.
  - Component with `Overrides` only → values from overrides.
  - Component with both → correct deep merge (precedence: file < overrides).
  - Component with nested maps in both → deep merge into the same key.
  - Missing values file path → `ErrCodeInternal` wrapping the read error.
- Add a bundler test asserting that `bundler.DefaultBundler.Make` and `RecipeResult.GetValuesForComponent` produce identical values for the same component — pins the equivalence and prevents future drift.

### Documentation

- Update `docs/contributor/data.md` to document the new public helpers and their merge order.
- Update the `pkg/recipe` package godoc.

## Acceptance

- `RecipeResult.GetValuesForComponent` and `recipe.GetManifestContent` are public, documented, and tested.
- `pkg/bundler` uses the new helpers internally — single code path for the merge.
- The bundler's existing behavior is byte-identical before and after the refactor (verify by running KWOK recipe tests against `main` and the PR branch, comparing emitted `values.yaml` for every component in every leaf).
- `make qualify` clean.
- `docs/contributor/data.md` updated.

## Out of scope

- Adding a `WithProvider` variant of either helper. That's a follow-up that pairs with #983's per-Builder DataProvider isolation work — once isolation lands, both helpers gain an optional `dp` parameter so in-process consumers with multiple recipe sources can bound the read to a specific provider. Until isolation lands, both helpers correctly fall back to the package-global `GetDataProvider()`.
- Changing the merge order. Pure refactor of where the algorithm lives, not what it does.
- Streaming or memory-bounded variants. The bundler already reads files into memory; this just exposes the same path.

## Dependencies

- **Soft dependency** on the per-Builder `DataProvider` isolation work (separate child issue). This issue can land first and a follow-up will add the `WithProvider` variants once isolation exists.
- **Hard prereq for #966's Prereq 1** (BOM-per-leaf rewrite). The BOM tool must consume these helpers, not duplicate the merge.

### PR DESCRIPTION
**Pull request by mchmarny**
## Summary

Adds per-Builder DataProvider isolation to `pkg/recipe` so it is safe for multi-tenant in-process consumers, plus provider-bound `GetValuesForComponent` / `GetManifestContent` variants. Marks process-global `SetDataProvider` / `GetDataProvider` as deprecated.

## Motivation / Context

`pkg/recipe` was single-tenant by design (concurrency was not required for CLI but plays role in external GO pkg usage and API server). The metadata store and component registry caches were keyed by a global generation counter incremented on `SetDataProvider`. Two `Builder`s against different providers in the same process could share state silently. This makes `pkg/recipe` usable as a library for operators, controllers, in-process testing harnesses, and BOM tooling.

Fixes: #985, #986, #987
Related: #983 (library-readiness epic), #966 (BOM tooling consumer)

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update
- [x] Refactoring (no functional changes)
- [ ] Build/CI/tooling

## Component(s) Affected

- [x] CLI (`cmd/aicr`, `pkg/cli`) — nolint comments on existing `SetDataProvider` call sites only
- [ ] API server (`cmd/aicrd`, `pkg/api`, `pkg/server`)
- [x] Recipe engine / data (`pkg/recipe`)
- [x] Bundlers (`pkg/bundler`, `pkg/component/*`) — parity test + `errors.AsType` modernization
- [ ] Collectors / snapshotter (`pkg/collector`, `pkg/snapshotter`)
- [x] Validator (`pkg/validator`) — `pkg/validator/catalog` nolint + `CutSuffix` refactor
- [ ] Core libraries (`pkg/errors`, `pkg/k8s`)
- [x] Docs/examples (`docs/`, `examples/`) — `docs/contributor/data.md`, `pkg/recipe/doc.go`

## Implementation Notes

- Replaced generation-counter caches with `sync.Map[DataProvider]*entry` keyed by provider identity. Per-entry `sync.Once` preserves populate-once semantics under concurrent load.
- `Builder.dp` flows to `MetadataStore.provider` to `RecipeResult.provider`, so `RecipeResult.GetValuesForComponent` reads from the same source that built the result.
- Added `EvictCachedStore` / `EvictCachedRegistry` for long-running consumers that retire providers.
- **SA1019 decision:** `nolint` at the call sites (`//nolint:staticcheck // tracked by #983 Stage 2`). gopls IDE warnings on the deprecated calls are inherent to Go's `Deprecated:` lifecycle and cannot be silenced separately from `golangci-lint` — accepted as the canonical mark-first-migrate-later pattern.
- Internal `pkg/recipe` calls were rewritten to consult the bound provider; only fallback branches (`if dp == nil { dp = GetDataProvider() }`) still reach the package-global.

## Testing

```bash
make qualify
```

New tests (all `-race` clean):

- `TestBundlerValueParity_WithRecipeResult` (pins bundler ↔ adapter equivalence)
- `TestNewBuilder_WithDataProvider` / `TestNewBuilder_DataProviderNilFallback`
- `TestLoadMetadataStore_PerProviderIsolation`
- `TestLoadMetadataStore_ConcurrentSameProviderIsCached`
- `TestEvictCachedStore_Refetches` / `_NilIsNoOp`
- `TestLoadMetadataStoreFor_NilProviderFallsBack`
- `TestGetComponentRegistry_PerProviderIsolation`
- `TestEvictCachedRegistry_Refetches` / `_NilIsNoOp`
- `TestGetComponentRegistryFor_NilProviderFallsBack`
- `TestBuilder_WithDataProvider_UsesBoundProvider`
- `TestBuilder_TwoProviders_ConcurrentBuild` (16 concurrent goroutines)
- `TestBuilder_DataProvider_NilSafe`
- `TestRecipeResult_GetValuesForComponent_HonorsBoundProvider`
- `TestGetManifestContentWithProvider` / `_NilFallback`
- `TestRecipeResult_GetValuesForComponent_BoundProviderSurvivesGlobalSwap`

**Coverage:** `pkg/recipe` 90.5% to 91.5% (+1.0%). All 8 new exported symbols have non-zero coverage. Project-wide test-coverage gate (75% floor): 76.9%. No touched-package regression > 0.5%.

## Risk Assessment

- [x] **Low** — Isolated change, well-tested, easy to revert
- [ ] **Medium** — Touches multiple components or has broader impact
- [ ] **High** — Breaking change, affects critical paths, or complex rollout

**Rollout notes:**

- **Cache memory:** monotonically grows with unique providers if callers never evict. In-tree CLI / API server create exactly one provider per process, so the back-compat path is bounded to one entry — same as before.
- **Behavior change:** none for CLI / API server (they don't use `WithDataProvider`). Bundler emits byte-identical bundles (pinned by `TestBundlerValueParity_WithRecipeResult`).
- **Deprecation:** godoc + lint markers only — no symbol removal, no runtime change. Stage 2 migration of in-tree call sites is tracked under #983.

## Checklist

- [x] Tests pass locally (`make test` with `-race`)
- [x] Linter passes (`make lint`)
- [x] I did not skip/disable tests to make CI green
- [x] I added/updated tests for new functionality
- [x] I updated docs if user-facing behavior changed
- [x] Changes follow existing patterns in the codebase
- [x] Commits are cryptographically signed (`git commit -S`)

### CODE REVIEW

File: docs/contributor/data.md [hunk 1]
```diff
@@ -1591,9 +1591,69 @@ Debug logs include:
...
 
-The data provider is initialized early in CLI command execution:
+`WithDataProvider` is the canonical way to attach a `DataProvider` to a recipe
+build. The returned `Builder` resolves its metadata store, component registry,
+and per-component values files through the bound provider — never through the
+process-global one. Each call site that builds recipes should construct its
+own provider and pass it through `WithDataProvider`:
+
+```go
+embedded := recipe.NewEmbeddedDataProvider(recipe.GetEmbeddedFS(), "data")
```
[the remark this answers is not in the document]
Comment (mchmarny) [reply docs/contributor/data.md]: Fixed in 52afac2a: changed prefix to empty string in both occurrences (lines 1603 and 1666). Verified line 1628 was already correct.

File: pkg/recipe/adapter.go [hunk 2]
```diff
@@ -30,11 +30,30 @@ func GetEmbeddedFS() embed.FS {
...
+// the global-provider fallback without an explicit nil check.
+//
+// Path should be relative to the data root (e.g.,
+// "components/network-operator/manifests/nfd-network-rule.yaml").
+func GetManifestContentWithProvider(dp DataProvider, path string) ([]byte, error) {
+	if dp == nil {
+		dp = GetDataProvider() //nolint:staticcheck // back-compat fallback for pre-WithDataProvider callers (#983 Stage 2)
+	}
+	return dp.ReadFile(path)
 }
```
[the remark this answers is not in the document]
Comment (mchmarny) [reply pkg/recipe/adapter.go:66]: Fixed in 52afac2a: GetManifestContentWithProvider now wraps with ErrCodeNotFound (on fs.ErrNotExist) / ErrCodeInternal (other failures) via pkg/errors.Wrap. The bundler caller's stderrors.Is(err, fs.ErrNotExist) chain is preserved by pkg/errors.Wrap Unwrap support — verified by new TestGetManifestContentWithProvider_NotFound and existing bundler tests.
